Fix: cloudsmith handling development repositories, release 0.3.0

n-dusan opened the pull request:
* Old script wasn't supporting specifying --version with `<,>~=`
Install package with version appended, so for example
something like `oll-test-top-level --version ~=0.1.0,<0.2.0` now works
* Support `cloudsmith_url` during upgrading a package, for example, now oll-test-top-level[development] correctly installs from cloudsmith.
* Add *args to install_with_constraints, so we can specify `--pre` and `--upgrade` flags for example
  * NOTE: `--pre` is only provided in the cli arguments if a `pip.conf` does not exist in venv! 
* Use `install_wheels` in `attempt_upgrade` instead of pip install. now constraints are used correctly for all upgrades
* Logging improvements. We can now output the end result of upgrade process with a `--format-output ` flag

TODO in a seperate PR: unit tests for response JSON ({"success", "responseError"})

n-dusan committed:
fix: update versions during non local install

* Old script wasn't supporting specifying --version with `<,>~=`
Install package with version appended, so for example
something like `oll-test-top-level~=0.1.0,<0.2.0` now works

n-dusan committed:
fix: support `cloudsmith_url` in upgrade, `--pre` flag

* Pip install using cloudsmith_url, if provided
* If `--pre` flag is provided via vars, use it to install from
  development cloudsmith

n-dusan committed:
fix: use `--extra-index-url` if cloudsmith is provided

* Black formatting

n-dusan committed:
fix,refact: call install_wheels inside `attempt_upgrade`

* For development repositories (packages which upgrade is attempted
  against) without --version specified, attempt to install them with
  constraints. Provide arbitrary *args to install_with_constraints.
* Formatting
* Refactor reinstalling with --no-deps

n-dusan committed:
feat: add output json

*If --format-output flag is provided, script returns a {success:[true|false],responseError: [string]} JSON response.

n-dusan committed:
fix: logging fixes

fix: use <py38 logger cleanup logic

* When printing out a result of upgrade process, we don't want to parse
  everything that happened during upgrade. This could potentially mean
  that we would parse a 1000 line txt
  document, where it would be difficult to determine if upgrade went
  through or not.
* Console scripts do not return string or object values, so we rely on
  printing out a JSON object.
* In order to make this easer, if `--format-output` flag is set, we
  clear logging handlers to enable DEBUG logging again, and print out
  the json response.
* This also helps us by having `--log-location` and `--format-output`
  flags set both at the same time.

renatav commented on upgrade/scripts/upgrade_python_package.py:
```diff
@@ -133,53 +155,61 @@ def install_wheel(package_name, cloudsmith_url=None, local=False, wheels_path=No
         # check if the wheel is already installed
-        wheel_metadata = pip("show", package_name.split("==")[0])
+        show_package = package_name.split("==")[0] if local else package_name
+        wheel_metadata = pip("show", show_package)
```
We don't have to implement this right away, but we can use `pip list`  to check if the package is already installed first. If the package is not installed, this show prints a very lengthy error which makes it difficult to analyze the output

n-dusan replied:
Okay, let's do that now

renatav commented on upgrade/scripts/upgrade_python_package.py:
```diff
@@ -455,14 +493,22 @@ def upgrade_python_package(
             cloudsmith_url,
             *vars,
         )
+    if format_output:
```
If an error happens when attempting to run the module, this code is never reached.

n-dusan replied:
Yes, thank you

renatav commented on upgrade/scripts/upgrade_python_package.py:
```diff
@@ -455,14 +493,22 @@ def upgrade_python_package(
         run_initial_post_install(package, *vars)
     else:
-        upgrade_and_run(
+        success, response_err = upgrade_and_run(
```
Ran
`py -m upgrade.scripts.upgrade_python_package oll-partners==2.6.0.dev16385 --cloudsmith-url=https://dl.cloudsmith.io/xxy/openlawlibrary/development/python/index/  --format-output --test
`
and that resulted in `ModuleNotFoundError: No module named 'oll_partners==2'` error. It will probably be necessary to update `try_running_module`

n-dusan replied:
I think we don't need to check if development version is provided with `==`. We can just do that with a `--version` flag

renatav commented on upgrade/scripts/upgrade_python_package.py:
```diff
@@ -204,36 +234,36 @@ def attempt_to_install_version(package_install_cmd, version, cloudsmith_url=None
     pip_config = pip("config", "list")
     pip_args = []
-    match = development_index_re.search(pip_config)
+    match = development_index_re.search(pip_config) or "--pre" in str(args)
```
I think that we also need to check cloudsmith url because this did not work `py -m upgrade.scripts.upgrade_python_package oll-partners --cloudsmith-url=https://dl.cloudsmith.io/xxy/openlawlibrary/development/python/index/  --format-output`

n-dusan replied:
Added

n-dusan committed:
refact: use `pip list` to check if package is already installed

* Based on #9 review, it makes sense to rework pip show to reduce error
  outputs. `pip show` returns an exit code, which can mess up our error
  handling. Parse pip list and see if there's a version matching it.

  For example, if package `attrs` exists in current venv, form a tuple (`attrs`, `21.6.2`)
  and return just _,version.

n-dusan committed:
refact: surround main with try except

* If a module fails, we want to print the output error

n-dusan committed:
feat: add check if cloudsmith url is valid

feat: add cloudsmith unit tests

* If CLOUDSMITH_URL variable is set, use it to test request to actual
  cloudsmith
